packages/shared-internals/src/dep-validation.ts

```diff
@@ -0,0 +1,108 @@
+import Package from './package';
+
+// For each package in the graph, discover all the paths to reach that package.
+// That is, we're identifying all its consumers.
+export function crawlDeps(startingPackage: Package): Map<Package, Package[][]> {
+  let queue: { pkg: Package; path: Package[] }[] = [{ pkg: startingPackage, path: [] }];
+  let seen = new Set<Package>();
+  let results = new Map<Package, Package[][]>();
+  for (;;) {
+    let entry = queue.shift();
+    if (!entry) {
+      break;
+    }
+    let { pkg, path } = entry;
+
+    let paths = results.get(pkg);
+    if (paths) {
+      paths.push(path);
+    } else {
+      results.set(pkg, [path]);
+    }
+
+    if (!seen.has(pkg)) {
+      seen.add(pkg);
+      for (let dep of pkg.dependencies) {
+        if (pkg.categorizeDependency(dep.name) !== 'peerDependencies') {
+          queue.push({ pkg: dep, path: [...path, pkg] });
+        }
+      }
+    }
+  }
+  return results;
+}
+
+interface PeerDepViolation {
+  // this package...
+  pkg: Package;
+  // sees this peer dep.
+  dep: Package;
+  // pkg was reached by this path of ancestors
+  ancestors: Package[];
+  // this particular ancestor...
+  ancestor: Package;
+  // provides a conflicting value for the peerDep
+  ancestorsDep: Package;
+}
+
+export function validatePeerDependencies(appPackage: Package): PeerDepViolation[] {
```

Comment:
is this pretty fast? 

academically, it's `O(N^4)`, but practically, are the iterations of the nested loops even that numerous? (like, ancestors and consumptions are both probably pretty small, yeah?)

Reply:
It's not really N^4 (where N is the number of packages), because those are different N's at the different levels. The outermost level is O(N) where N=number of packages. But the next level is O(P) where P = average number of peer deps a package has. Which is probably closer to zero than one. That immediately chops off a vast swath of the space.

Then if you do have peer deps, the next level is the number of times your package is consumed, and the final level is the number of level we have to look upward to find the provider of the peer dep.

It's quite fast, and it's all traversing our in-memory package cache.